### ISSUE TITLE
sql: support NULLS FIRST/LAST order by modifiers

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#6224]
Many SQL engines (including Oracle and Postgres) support the `ORDER BY` modifiers of `NULLS {FIRST | LAST}`. These modifiers, as their names' suggest, specify that NULL values should be returned before/after non-NULL values. We should consider if we want to support this syntax given that it could be pretty useful for queries on any sparse tables. If so, it might take a bit of work to make sure both modifiers work correctly with ordering which uses an index to avoid in-memory sorting.

**Comment by petermattis** [comment]
Postgres also allows you to specify whether NULLS sort first or last for indexes. This sort of flexibility seems a bit overkill to me, though. Do you have a specific example where it would be useful?

**Comment by nvanbenschoten** [comment]
Like I touched on above, you could imagine a spare table where a user is relying on NULL values for the majority of elements, but then has a query like "what are the top x values _that exist_". Granted an identical behavior could be found by using a `CASE` statement within the `ORDER BY` clause, but `NULLS {FIRST | LAST}` would be a more expressive way to achieve this.

And the only reason I even noticed the need for this is that it seems like our default (only) behavior is the opposite of Postgres' (but in-line with the SQL standard).

**Comment by petermattis** [comment]
`SELECT x FROM t WHERE x IS NOT NULL ORDER BY x` will construct a scan that skips over the `NULL` values of `x` (assuming there is an index on `x`). Similarly, `SELECT x FROM t WHERE x IS NOT NULL ORDER BY x DESC` will do a reverse scan, stopping before it gets to the `NULL` values. Controlling whether `NULLs` sort first or last feels esoteric to me, but I'd be happy to be educated on why the flexibility is important.

**Comment by bdarnell** [comment]
I think the use case is for something like a sortable table: if I sort by last name, I expect to see last names A-Z before any records without last names. If I reverse that sort option, I (arguably) want to see last names Z-A before any records without a name, instead of an exact reversal of the previous order. This can be achieved with `ORDER BY last_name NULLS LAST` and `ORDER BY last_name DESC NULLS LAST`. Note that `NULLS FIRST` and `NULLS LAST` are independent of the `ASC`/`DESC` ordering used for non-null values.

**Comment by nvanbenschoten** [comment]
I think Ben's example is pretty good, as it demonstrates that instead of filtering records without last names, a user might just want to control where they end up in the sort order. Again, this could be handled with a query like `SELECT x FROM t ORDER BY CASE  WHEN x IS NULL THEN 1 ELSE 0 END, x`, but this seems a little verbose.

On a side note, it looks like the default sort order of NULL values is left undefined by the SQL spec. We seem to order them first by default (along with a few other databases), while Postgres seems to order them last. Does this seem ok to you two? Would we prefer to follow PG on unspecified behavior like this?

**Comment by bdarnell** [comment]
The `ORDER BY CASE` is not only verbose, it seems much harder for index selection to make sense of than an explicit `NULLS FIRST/LAST` option. 

I have a very slight personal preference for nulls-last, but not enough to change the way indexes are encoded. According to the chart at the bottom of http://use-the-index-luke.com/sql/sorting-grouping/order-by-asc-desc-nulls-last, oracle and postgresql default to nulls-last while db2, mysql, sqlite and ms sql are nulls-first. Both databases that default to nulls-last support changing this with the `NULLS FIRST` clause, while in all of the ones that default to nulls-first this cannot be changed. So it looks like we'll be in good company if we are inflexibly nulls-first.

**Comment by petermattis** [comment]
Other than the plumbing required, it looks like supports `NULLS LAST` for indexing wouldn't be that hard (we already have `EncodeNullDescending` which encodes a value greater than any other encoded value). That said, I feel @bdarnell's use case is contrived. Good to have this issue, but let's leave implementation until there is user need or someone has spare time.

**Comment by retr0h** [comment]
I've been trying to port some code from pg to cockroach.

The following query fails.

```
PREPARE pq_XAFN5SIPGWGXOQTTHV6YOLOKG2PPEZRRAQ2C7RIQNVOKBHMDKAQQ AS
    WITH
      selected AS (
        SELECT * FROM queue
        WHERE
          q_name = 'apples' AND
          dequeued_at IS NULL
        ORDER BY schedule_at nulls first, expected_at nulls last, id
```

```
root@:26257/fibre> select * from queue WHERE q_name = 'apples';
          id         |           enqueued_at            | dequeued_at | expected_at | schedule_at | q_name |             data
---------------------+----------------------------------+-------------+-------------+-------------+--------+-------------------------------
  569570299516518401 | 2020-07-04 19:04:21.124729+00:00 | NULL        | NULL        | NULL        | apples | {"kind": "Cox"}
  569570299538604033 | 2020-07-04 19:04:21.137027+00:00 | NULL        | NULL        | NULL        | apples | {"kind": "Arthur Turner"}
  569570299555905537 | 2020-07-04 19:04:21.14234+00:00  | NULL        | NULL        | NULL        | apples | {"kind": "Golden Delicious"}
  569570517121826817 | 2020-07-04 19:05:27.537352+00:00 | NULL        | NULL        | NULL        | apples | {"kind": "Cox"}
  569570517139783681 | 2020-07-04 19:05:27.543881+00:00 | NULL        | NULL        | NULL        | apples | {"kind": "Arthur Turner"}
  569570517155971073 | 2020-07-04 19:05:27.54866+00:00  | NULL        | NULL        | NULL        | apples | {"kind": "Golden Delicious"}
(6 rows)
```

I'm trying to wrap my brain around the alternative suggestions above, but I'm not having much luck understanding it.  Would someone mind helping me with an appropriate query?

**Comment by nvanbenschoten** [comment]
The workaround discussed above would lead to a query that looks like:
```sql
PREPARE pq_XAFN5SIPGWGXOQTTHV6YOLOKG2PPEZRRAQ2C7RIQNVOKBHMDKAQQ AS
    WITH
      selected AS (
        SELECT * FROM queue
        WHERE
          q_name = 'apples' AND
          dequeued_at IS NULL
        ORDER BY
          schedule_at,
          CASE WHEN expected_at IS NULL THEN 1 ELSE 0 END,
          expected_at,
          id
```
That should work to get you around this issue, though you'll want to be careful that it doesn't lead to an overly inefficient query. Since the primary ordering column was already sorting by nulls first, this probably won't cause any issues with index selection, but it will require more work during sorting. For small result sets, this is likely negligible.

**Comment by tobiasmuehl** [comment]
Using this feature in Postgres a lot, with tables that use a timestamp column instead of a boolean column. For example, a `user`s table that has a nullable timestamp column `email_verified`. I care about the unverified users, so I build an index on `email_verified NULLS FIRST` that can efficiently be used in a query `SELECT x FROM user WHERE email_verified IS NULL`

**Comment by bdarnell** [comment]
You don't need NULLS FIRST for that. `WHERE email_verified IS NULL` will find the rows with NULLs efficiently, whether they're at the beginning or end of the index. You only need `NULLS FIRST/LAST` for orderings when you are mixing null and non-null values.

**Comment by ajwerner** [comment]
FWIW that above example is super silly because, well, primary key constraints imply `NOT NULL`. I mean to say, for this case:

```
root@localhost:26257/defaultdb> create table foo (i INT, j INT, INDEX (j ASC NULLS LAST));
invalid syntax: statement ignored: at or near "last": syntax error: unimplemented: this syntax
SQLSTATE: 0A000
DETAIL: source SQL:
create table foo (i INT, j INT, INDEX (j ASC NULLS LAST))
                                                   ^
HINT: You have attempted to use a feature that is not yet implemented.
See: https://go.crdb.dev/issue-v/6224/v21.2
```

**Comment by nehageorge** [comment]
I was going to close this issue and open a new one for index modifiers, but I realized that the unimplemented syntax error links to this issue. Updated the description instead to reflect the work that has been done.

**Comment by vy-ton** [comment]
> However, this issue also requires that the modifiers be added for indexes.

@jordanlewis @rytaft Could we complete this as a starter project?

**Comment by rytaft** [comment]
I think this would be pretty involved, requiring changes in both schema and the optimizer, and possibly even lower in the stack (seems like we would need to add a new encoding to support nulls last?). Maybe there's an easier way that I'm not thinking of, but I'd be hesitant to give this to someone brand new (at least not as their first or second starter project...). It might be doable as the last starter project.

**Comment by jordanlewis** [comment]
I agree, it's not trivial. For the encoding I *think* we can reuse the NULL DESCENDING byte for encoding nulls in ascending NULLS LAST indexes, and the NULL ASCENDING byte for encoding nulls in descending NULLS FIRST indexes. But that doesn't make it too much easier!

**Comment by rytaft** [comment]
This is not being actively worked on, and it's currently not in the plan for 22.2. It was originally one of @msirek's starter projects, but it got deprioritized in favor of other projects for 22.1.

**Comment by BramGruneir** [comment]
Any way to get this into 22.2?  Or is it too large of a project?

**Comment by vy-ton** [comment]
Based on the comments above, it's a non-trivial project, so we will need to re-evaluate during 23.x planning.

> To reiterate, NULLS FIRST/LAST modifiers were added for ORDER BY in this PR https://github.com/cockroachdb/cockroach/pull/71083. However, this issue also requires that the modifiers be added for indexes.

@BramGruneir Just want to confirm that you need the modifiers for indexes?

**Comment by BramGruneir** [comment]
Yes, that was the request.  They are using it in Postgres and wanted the same functionality.

**Comment by awoods187** [comment]
This came up recently in unimplemented telemetry

**Comment by giangpham712** [comment]
@fqazi This affects efcore.pg tests

**Comment by fqazi** [comment]
@giangpham712 Let's skip the efcore test that need this for now
